### ISSUE TITLE
Clarify that metrics are omitted when their value is not yet determined or unknown

### ISSUE THREAD

**Issue by henbos** [w3c/webrtc-stats#736]
"4.2 Guidelines for implementing stats objects" already says to use undefined if "a feature is not applicable" and gives the example of an audio metric for a video stream. We may want to clarify that if a metric is not yet determined or not yet known then we should also report undefined.

For example, a video stream will have an encoderImplementation, but which one will be used is not determined until after the encoder has been initialized.

Chrome currently reports the string value "unknown" but this should be undefined. There is no encoderImplementation with the name "unknown" and we have not spec'ed this magic value so it shouldn't exist

**Comment by henbos** [comment]
@fippo

**Comment by henbos** [comment]
I don't think this is controversial so using labels "Editorial" and "Ready for PR"

**Comment by fippo** [comment]
Yeah, this is mostly wordsmithing. But behold, this is a can of worms.

IMO encoderImplementation/decoderImplementation fall under this
```
When a counted feature hasn't been used yet, but may happen in the future, report a count of zero.
```
but for strings. For these it may make sense to include the empty string despite guidance against this in the sentence before.
I *think* it is the only two cases? Easy... but behold...

"may happen in the future" is vague. What this means is you:
* should include packetsLost (even though you may never loose a packet)
since it means that you can safely subtract stats from two times T_a and T_b without worrying about a value being undefined a T_a.

This applies to some things that are not "counted" but should be covered like [STUN](https://w3c.github.io/webrtc-stats/#dom-rtcicecandidatepairstats-totalroundtriptime) RTT. This should be 0 initially and be present. It is clear that this has not happened yet because responsesReceived is still 0.

STUN currentRoundTripTime however should not be present before a response has been received.
Neither should audioLevel before you decode or send a packet.
Both values are instantaneous. Following that logic, decoderImplementation should not be set either.

For RTCP we have conflicting guidance in [here](https://w3c.github.io/webrtc-stats/#dictionary-rtcremoteinboundrtpstreamstats-members) to only include when a RR with a non-zero DLSR field was received (and yep, that takes some time, say hello to flaky tests). This seems slightly off, the intention is to count valid round trip times and only increment for those.


Now trying to summarize that I would say the guidance should be
* When a cumulative feature hasn't been used yet report a count of zero so stats values from two points in time can be safely subtracted.
* When an momentary feature hasn't been used yet, omit the dictionary member

**Comment by henbos** [comment]
If there is no current (e.g. currentRoundTripTime), then it is not applicable. Not applicable => undefined.
Same reasoning applies to encoderImplementation. If there is no current encoder, then enoderImplementation is not applicable, so its undefined.

A counted feature starting at 0 is separate, if something has happened no times then the counter is 0, its a superflous sentence in the spec

**Comment by jan-ivar** [comment]
> "4.2 Guidelines for implementing stats objects" already says to use undefined if "a feature is not applicable"

It doesn't say that. It says: _"When a feature is not applicable to an instance of an object (for example [audioLevel](https://w3c.github.io/webrtc-stats/#dom-rtcinboundrtpstreamstats-audiolevel) on a video stream), **omit** the dictionary member. Do NOT report a count of zero, -1 or "empty string"."_

It says to _"omit"_ the member, which is different from a value of `undefined`. [audioLevel](https://w3c.github.io/webrtc-stats/#dom-rtcinboundrtpstreamstats-audiolevel): _"Only [exists](https://infra.spec.whatwg.org/#map-exists) for audio."_

E.g. here `audioLevel` [exists](https://infra.spec.whatwg.org/#map-exists):
```js
{audioLevel: undefined}
```
...and here it is omitted:
```js
{}
```
AFAIK WebIDL bindings coerce the former into the latter on _input_ (e.g. dictionaries as arguments to a method), but not output.

**Comment by jan-ivar** [comment]
_(I'm interpreting defined/undefined in this thread has missing/present/[exists](https://infra.spec.whatwg.org/exists). See https://github.com/w3c/webrtc-stats/issues/737. LMK if that's not right)_

While there might be some value today in knowing immediately whether a browser supports a stat or not, such value should diminish over time, so omitting members until they can be accurately reported seems superior to me to the alternatives.

There's probably some precedence here with stats objects of type [remote-inbound-rtp](https://w3c.github.io/webrtc-stats/#dom-rtcstatstype-remote-inbound-rtp) and [remote-outbound-rtp](https://w3c.github.io/webrtc-stats/#dom-rtcstatstype-remote-outbound-rtp) which don't appear until rtcp packets have been received (comment out the `await wait(5000)` in [this fiddle](https://jsfiddle.net/jib1/moz71L7h/).

**Comment by henbos** [comment]
You're right I meant missing when I said undefined. We should fix #737.
